src/neo/Network/P2P/Payloads/TransactionAttributeCollection.cs

```diff
@@ -0,0 +1,125 @@
+using Neo.IO;
+using Neo.IO.Json;
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+
+namespace Neo.Network.P2P.Payloads
+{
+    public class TransactionAttributeCollection : ISerializable, IEnumerable<KeyValuePair<TransactionAttributeUsage, TransactionAttribute>>
+    {
+        /// <summary>
+        /// Maximum number of attributes that can be contained within a transaction
+        /// </summary>
+        public const int MaxTransactionAttributes = 16;
+
+        private readonly Dictionary<TransactionAttributeUsage, TransactionAttribute> _entries;
+
+        public int Size =>
+            IO.Helper.GetVarSize(_entries.Count) +  // count
+            _entries.Count +                        // usages
+            _entries.Values.Sum(u => u.Size);       // entries
+
+        public int Count => _entries.Count;
+
+        public TransactionAttribute this[TransactionAttributeUsage usage]
+        {
+            get
+            {
+                if (_entries.TryGetValue(usage, out var attr))
+                {
+                    return attr;
+                }
+
+                return null;
+            }
+        }
+
+        public TransactionAttributeCollection()
+        {
+            _entries = new Dictionary<TransactionAttributeUsage, TransactionAttribute>();
+        }
+
+        public TransactionAttributeCollection(params TransactionAttribute[] attributes)
+        {
+            _entries = new Dictionary<TransactionAttributeUsage, TransactionAttribute>();
+
+            if (attributes != null)
+            {
+                foreach (var attr in attributes)
+                {
+                    Add(attr.Usage, attr);
+                }
+            }
+        }
+
+        public void Serialize(BinaryWriter writer)
+        {
+            writer.WriteVarInt(_entries.Count);
+            foreach (var attr in _entries)
+            {
+                writer.Write((byte)attr.Key);
+                writer.Write(attr.Value);
+            }
+        }
+
+        public void Deserialize(BinaryReader reader)
+        {
+            _entries.Clear();
+
+            var count = (int)reader.ReadVarInt(MaxTransactionAttributes);
+            for (int x = 0; x < count; x++)
+            {
+                switch ((TransactionAttributeUsage)reader.ReadByte())
+                {
+                    case TransactionAttributeUsage.Cosigner:
+                        {
+                            _entries.Add(TransactionAttributeUsage.Cosigner, reader.ReadSerializable<CosignerAttribute>());
```

Comment:
I'm not sure, we only can have one attribute per type? Like `OracleReponseAttribute`, I think one oracle response transaction can have multiple attributes.

Reply:
I was thinking about it and I think that it's better and faster only one per type. We can create a collection of it like the consigners.